presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownTableConstraints.java

```diff
@@ -0,0 +1,141 @@
+        Expression deterministicPredicate = stripNonDeterministicConjuncts(predicate);
+
+        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+                metadata,
+                session,
+                deterministicPredicate,
+                symbolAllocator.getTypes());
+
+        TableScanNode tableScan = (TableScanNode) lookup.resolve(filter.getSource());
+        TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
+                .transform(tableScan.getAssignments()::get)
+                .intersect(tableScan.getCurrentConstraint());
+
+        Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(tableScan.getAssignments()).inverse();
+
+        // Layouts will be returned in order of the connector's preference
+        List<TableLayoutResult> layouts = metadata.getLayouts(
+                session, tableScan.getTable(),
+                new Constraint<>(simplifiedConstraint, bindings -> true),
+                Optional.of(tableScan.getOutputSymbols().stream()
+                        .map(tableScan.getAssignments()::get)
+                        .collect(toImmutableSet())));
+
+        if (layouts.isEmpty()) {
+            return Optional.of(new ValuesNode(idAllocator.getNextId(), tableScan.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        // Filter out layouts that cannot supply all the required columns
+        layouts = layouts.stream()
+                .filter(layoutHasAllNeededOutputs(tableScan))
+                .collect(toList());
+        checkState(!layouts.isEmpty(), "No usable layouts for %s", tableScan);
+
+        List<PlanNode> possiblePlans = layouts.stream()
+                .map(layout -> {
+                    TableScanNode rewrittenTableScan = new TableScanNode(
+                            tableScan.getId(),
+                            tableScan.getTable(),
+                            tableScan.getOutputSymbols(),
+                            tableScan.getAssignments(),
+                            Optional.of(layout.getLayout().getHandle()),
+                            simplifiedConstraint.intersect(layout.getLayout().getPredicate()),
+                            Optional.ofNullable(tableScan.getOriginalConstraint()).orElse(predicate));
+
+                    Expression resultingPredicate = combineConjuncts(
+                            DomainTranslator.toPredicate(layout.getUnenforcedConstraint().transform(assignments::get)),
+                            stripDeterministicConjuncts(predicate),
+                            decomposedPredicate.getRemainingExpression());
+
+                    if (!BooleanLiteral.TRUE_LITERAL.equals(resultingPredicate)) {
+                        return new FilterNode(idAllocator.getNextId(), rewrittenTableScan, resultingPredicate);
+                    }
+
+                    return rewrittenTableScan;
+                })
+                .collect(toList());
```

Comment:
why operate on stream if later on you only consider first element? 
Maybe just operate on `layouts.get(0)`